File: backend/app/services/checksum.py
"""Structural / checksum validation.

+---------------------------------------------------------------------------+
| NOT AI. Every function in this module is deterministic arithmetic or a     |
| regex. There is no model, no training, no inference. UI copy for these     |
| results must say "deterministic check", never "AI".                        |
+---------------------------------------------------------------------------+

Covers:
  * Aadhaar  -- Verhoeff checksum (UIDAI spec)
  * Passport -- ICAO 9303 MRZ check digits (TD3)
  * Driving Licence -- structural format + state-code whitelist ONLY.
    No public national checksum standard exists for Indian DL numbers, so we
    do not invent one.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, asdict
from typing import Any

# --------------------------------------------------------------------------- #
# Result type
# --------------------------------------------------------------------------- #


@dataclass
class CheckResult:
    check: str
    passed: bool | None  # None == could not be evaluated (missing input)
    detail: str
    is_ai: bool = False  # always False in this module, kept explicit for the UI

    def dict(self) -> dict[str, Any]:
        return asdict(self)


# --------------------------------------------------------------------------- #
# Verhoeff (Aadhaar)
# --------------------------------------------------------------------------- #

# Dihedral group D5 multiplication table
_D = (
    (0, 1, 2, 3, 4, 5, 6, 7, 8, 9),
    (1, 2, 3, 4, 0, 6, 7, 8, 9, 5),
    (2, 3, 4, 0, 1, 7, 8, 9, 5, 6),
    (3, 4, 0, 1, 2, 8, 9, 5, 6, 7),
    (4, 0, 1, 2, 3, 9, 5, 6, 7, 8),
    (5, 9, 8, 7, 6, 0, 4, 3, 2, 1),
    (6, 5, 9, 8, 7, 1, 0, 4, 3, 2),
    (7, 6, 5, 9, 8, 2, 1, 0, 4, 3),
    (8, 7, 6, 5, 9, 3, 2, 1, 0, 4),
    (9, 8, 7, 6, 5, 4, 3, 2, 1, 0),
)

# Permutation table
_P = (
    (0, 1, 2, 3, 4, 5, 6, 7, 8, 9),
    (1, 5, 7, 6, 2, 8, 3, 0, 9, 4),
    (5, 8, 0, 3, 7, 9, 6, 1, 4, 2),
    (8, 9, 1, 6, 0, 4, 3, 5, 2, 7),
    (9, 4, 5, 3, 1, 2, 6, 8, 7, 0),
    (4, 2, 8, 6, 5, 7, 3, 9, 0, 1),
    (2, 7, 9, 3, 8, 0, 6, 4, 1, 5),
    (7, 0, 4, 6, 9, 1, 3, 2, 5, 8),
)

# Multiplicative inverse table (used only to *generate* valid test fixtures)
_INV = (0, 4, 3, 2, 1, 5, 6, 7, 8, 9)


def verhoeff_valid(number: str) -> bool:
    """True if `number` (digits only) carries a valid Verhoeff check digit."""
    digits = re.sub(r"\D", "", number or "")
    if not digits:
        return False
    c = 0
    for i, ch in enumerate(reversed(digits)):
        c = _D[c][_P[i % 8][int(ch)]]
    return c == 0


def verhoeff_check_digit(payload: str) -> int:
    """Check digit that would make `payload` + digit a valid Verhoeff number."""
    digits = re.sub(r"\D", "", payload or "")
    c = 0
    for i, ch in enumerate(reversed(digits)):
        c = _D[c][_P[(i + 1) % 8][int(ch)]]
    return _INV[c]


AADHAAR_RE = re.compile(r"\b(\d{4})\s?(\d{4})\s?(\d{4})\b")


def validate_aadhaar(number: str | None) -> list[CheckResult]:
    results: list[CheckResult] = []
    digits = re.sub(r"\D", "", number or "")

    if not digits:
        return [
            CheckResult(
                "aadhaar_present",
                None,
                "No 12-digit Aadhaar number was found in the extracted text.",
            )
        ]

    ok_len = len(digits) == 12
    results.append(
        CheckResult(
            "aadhaar_format",
            ok_len,
            "12 digits present."
            if ok_len
            else f"Expected 12 digits, found {len(digits)}.",
        )
    )

    # UIDAI never issues numbers beginning with 0 or 1.
    ok_lead = digits[0] not in "01"
    results.append(
        CheckResult(
            "aadhaar_leading_digit",
            ok_lead,
            "Leading digit is in the issued range (2-9)."
            if ok_lead
            else f"Aadhaar numbers never start with {digits[0]}.",
        )
    )

    if ok_len:
        ok_sum = verhoeff_valid(digits)
        results.append(
            CheckResult(
                "aadhaar_verhoeff_checksum",
                ok_sum,
                "Verhoeff checksum valid (deterministic arithmetic, not AI)."
                if ok_sum
                else "Verhoeff checksum failed — the number cannot be a genuine "
                "Aadhaar number as printed.",
            )
        )
    return results


# --------------------------------------------------------------------------- #
# ICAO 9303 MRZ (Passport, TD3)
# --------------------------------------------------------------------------- #

_WEIGHTS = (7, 3, 1)


def mrz_char_value(ch: str) -> int:
    if ch == "<":
        return 0
    if ch.isdigit():
        return int(ch)
    if "A" <= ch.upper() <= "Z":
        return ord(ch.upper()) - 55  # A=10 ... Z=35
    return 0


def mrz_check_digit(field: str) -> int:
    total = 0
    for i, ch in enumerate(field):
        total += mrz_char_value(ch) * _WEIGHTS[i % 3]
    return total % 10


def _digit_check(name: str, field: str, given: str, label: str) -> CheckResult:
    if not given.isdigit():
        return CheckResult(name, None, f"{label}: check digit is missing or unreadable.")
    expected = mrz_check_digit(field)
    ok = expected == int(given)
    return CheckResult(
        name,
        ok,
        f"{label}: check digit matches (ICAO 9303, deterministic)."
        if ok
        else f"{label}: check digit is {given}, expected {expected}.",
    )


def validate_passport_mrz(
    line1: str | None, line2: str | None, visual_passport_number: str | None = None
) -> list[CheckResult]:
    """Validate the four check digits of a TD3 machine-readable zone."""
    results: list[CheckResult] = []

    if not line2:
        return [
            CheckResult(
                "mrz_present",
                None,
                "No machine-readable zone was found. Capture the bottom two "
                "lines of the passport data page and retry.",
            )
        ]

    l2 = line2.strip().replace(" ", "").upper()
    ok_len = len(l2) == 44
    results.append(
        CheckResult(
            "mrz_line_length",
            ok_len,
            "MRZ line 2 is the expected 44 characters."
            if ok_len
            else f"MRZ line 2 is {len(l2)} characters, expected 44 — OCR may have "
            "dropped characters.",
        )
    )
    if not ok_len:
        return results

    if line1:
        l1 = line1.strip().replace(" ", "").upper()
        results.append(
            CheckResult(
                "mrz_document_type",
                l1.startswith("P"),
                "Document type is P (passport)."
                if l1.startswith("P")
                else f"Document type code is '{l1[:1]}', expected 'P'.",
            )
        )

    doc_no, doc_cd = l2[0:9], l2[9]
    dob, dob_cd = l2[13:19], l2[19]
    exp, exp_cd = l2[21:27], l2[27]
    personal = l2[28:42]
    composite_cd = l2[43]

    results.append(_digit_check("mrz_document_number", doc_no, doc_cd, "Passport number"))
    results.append(_digit_check("mrz_date_of_birth", dob, dob_cd, "Date of birth"))
    results.append(_digit_check("mrz_expiry", exp, exp_cd, "Expiry date"))

    composite = doc_no + doc_cd + dob + dob_cd + exp + exp_cd + personal + l2[42]
    results.append(
        _digit_check("mrz_composite", composite, composite_cd, "Composite (whole MRZ)")
    )

    if visual_passport_number:
        clean_mrz = doc_no.replace("<", "").strip().upper()
        clean_vis = re.sub(r"[^A-Z0-9]", "", visual_passport_number).upper()
        match = clean_mrz == clean_vis
        results.append(
            CheckResult(
                "passport_mrz_vs_visual_number",
                match,
                "MRZ passport number matches the visual zone number."
                if match
                else f"MRZ passport number ({clean_mrz}) conflicts with visual zone ({clean_vis}).",
            )
        )
    return results


# --------------------------------------------------------------------------- #
# Driving Licence -- structure only
# --------------------------------------------------------------------------- #

STATE_CODES = {
    "AP", "AR", "AS", "BR", "CG", "GA", "GJ", "HR", "HP", "JH", "KA", "KL",
    "MP", "MH", "MN", "ML", "MZ", "NL", "OD", "PB", "RJ", "SK", "TN", "TS",
    "TR", "UP", "UK", "WB", "AN", "CH", "DN", "DD", "DL", "JK", "LA", "LD",
    "PY",
}

DL_RE = re.compile(r"^[A-Z]{2}\d{2}\d{4,11}$")

DL_DISCLAIMER = (
    "No public national checksum standard exists for Indian driving licence "
    "numbers — structural check only. A pass here does not prove the licence "
    "is genuine."
)


def validate_dl(number: str | None) -> list[CheckResult]:
    results: list[CheckResult] = []
    raw = re.sub(r"[\s\-]", "", (number or "")).upper()

    if not raw:
        return [
            CheckResult(
                "dl_present",
                None,
                "No driving licence number was found in the extracted text.",
            )
        ]

    ok_fmt = bool(DL_RE.match(raw))
    results.append(
        CheckResult(
            "dl_format",
            ok_fmt,
            "Matches the state + RTO + serial layout."
            if ok_fmt
            else "Does not match the expected layout (2 letters, 2-digit RTO "
            "code, then 4-11 digits).",
        )
    )

    state = raw[:2]
    ok_state = state in STATE_CODES
    results.append(
        CheckResult(
            "dl_state_code",
            ok_state,
            f"'{state}' is a recognised Indian state / UT code."
            if ok_state
            else f"'{state}' is not a recognised Indian state / UT code.",
        )
    )

    results.append(CheckResult("dl_checksum_unavailable", None, DL_DISCLAIMER))
    return results


def validate_consistency(fields: dict[str, Any]) -> list[CheckResult]:
    results: list[CheckResult] = []

    dob_str = fields.get("dob") or fields.get("date_of_birth")
    issue_str = fields.get("issue_date") or fields.get("date_of_issue")
    expiry_str = fields.get("expiry_date") or fields.get("date_of_expiry") or fields.get("valid_until")

    def _parse_year(d: Any) -> int | None:
        if not d or not isinstance(d, str):
            return None
        m = re.search(r"\b(19\d\d|20\d\d)\b", d)
        return int(m.group(1)) if m else None

    dob_y = _parse_year(dob_str)
    issue_y = _parse_year(issue_str)
    expiry_y = _parse_year(expiry_str)

    if dob_y and issue_y:
        ok = dob_y < issue_y
        results.append(
            CheckResult(
                "consistency_dob_vs_issue",
                ok,
                f"Date of birth year ({dob_y}) precedes issue year ({issue_y})."
                if ok
                else f"Date of birth year ({dob_y}) cannot be after or equal to issue year ({issue_y}).",
            )
        )

    if issue_y and expiry_y:
        ok = issue_y < expiry_y
        results.append(
            CheckResult(
                "consistency_issue_vs_expiry",
                ok,
                f"Issue year ({issue_y}) precedes expiry year ({expiry_y})."
                if ok
                else f"Expiry year ({expiry_y}) precedes or equals issue year ({issue_y}).",
            )
        )

    return results


def validate(document_type: str, fields: dict[str, Any]) -> list[dict[str, Any]]:
    dt = (document_type or "").lower()
    if dt == "aadhaar":
        checks = validate_aadhaar(fields.get("aadhaar_number"))
    elif dt == "passport":
        checks = validate_passport_mrz(
            fields.get("mrz_line1"),
            fields.get("mrz_line2"),
            fields.get("visual_passport_number"),
        )
    elif dt in ("dl", "driving_licence", "driving_license"):
        checks = validate_dl(fields.get("dl_number"))
    else:
        checks = [
            CheckResult(
                "unsupported_document_type",
                None,
                f"'{document_type}' is not a supported document type. "
                "Supported: aadhaar, passport, dl.",
            )
        ]

    checks.extend(validate_consistency(fields))
    return [c.dict() for c in checks]


def summarise(checks: list[dict[str, Any]]) -> dict[str, Any]:
    """Plain-English rollup. Never a bare colour or number in the UI."""
    failed = [c for c in checks if c["passed"] is False]
    passed = [c for c in checks if c["passed"] is True]
    unknown = [c for c in checks if c["passed"] is None]

    if failed:
        status = "fail"
        headline = f"{len(failed)} structural check(s) failed."
    elif not passed:
        status = "needs_review"
        headline = "No structural check could be evaluated — recapture the document."
    elif unknown:
        status = "needs_review"
        headline = f"{len(passed)} check(s) passed, {len(unknown)} could not be evaluated."
    else:
        status = "pass"
        headline = f"All {len(passed)} structural checks passed."

    return {
        "status": status,
        "headline": headline,
        "passed": len(passed),
        "failed": len(failed),
        "not_evaluated": len(unknown),
        "reasons": [c["detail"] for c in failed + unknown] or [c["detail"] for c in passed],
        "method": "deterministic",
        "is_ai": False,
    }

File: src/lib/engine/checksum.ts
/**
 * Deterministic structural validation.
 *
 * Nothing in this file is AI. It is published check-digit arithmetic
 * (Verhoeff for Aadhaar, ICAO 9303 for passports) plus format grammar for
 * driving licences. A pass proves a number is well-formed — never that the
 * document or the holder is genuine.
 */

export type CheckItem = {
  check: string;
  label: string;
  passed: boolean | null;
  detail: string;
  kind: "deterministic" | "ai";
};

const D = [
  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
  [1, 2, 3, 4, 0, 6, 7, 8, 9, 5],
  [2, 3, 4, 0, 1, 7, 8, 9, 5, 6],
  [3, 4, 0, 1, 2, 8, 9, 5, 6, 7],
  [4, 0, 1, 2, 3, 9, 5, 6, 7, 8],
  [5, 9, 8, 7, 6, 0, 4, 3, 2, 1],
  [6, 5, 9, 8, 7, 1, 0, 4, 3, 2],
  [7, 6, 5, 9, 8, 2, 1, 0, 4, 3],
  [8, 7, 6, 5, 9, 3, 2, 1, 0, 4],
  [9, 8, 7, 6, 5, 4, 3, 2, 1, 0],
];

const P = [
  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
  [1, 5, 7, 6, 2, 8, 3, 0, 9, 4],
  [5, 8, 0, 3, 7, 9, 6, 1, 4, 2],
  [8, 9, 1, 6, 0, 4, 3, 5, 2, 7],
  [9, 4, 5, 3, 1, 2, 6, 8, 7, 0],
  [4, 2, 8, 6, 5, 7, 3, 9, 0, 1],
  [2, 7, 9, 3, 8, 0, 6, 4, 1, 5],
  [7, 0, 4, 6, 9, 1, 3, 2, 5, 8],
];

/** Verhoeff checksum — the scheme UIDAI uses for the 12-digit Aadhaar number. */
export function verhoeffValid(digits: string): boolean {
  if (!/^\d+$/.test(digits)) return false;
  let c = 0;
  const reversed = digits.split("").reverse();
  reversed.forEach((ch, i) => {
    const row = D[c];
    const permRow = P[i % 8];
    if (!row || !permRow) return;
    c = row[permRow[Number(ch)] ?? 0] ?? 0;
  });
  return c === 0;
}

/** ICAO 9303 check digit over an MRZ field (weights 7-3-1). */
export function icaoCheckDigit(field: string): number {
  const weights = [7, 3, 1];
  let sum = 0;
  for (let i = 0; i < field.length; i += 1) {
    const ch = field[i] ?? "";
    let value: number;
    if (ch >= "0" && ch <= "9") value = Number(ch);
    else if (ch >= "A" && ch <= "Z") value = ch.charCodeAt(0) - 55;
    else value = 0; // '<' filler
    sum += value * (weights[i % 3] ?? 1);
  }
  return sum % 10;
}

export const icaoValid = (field: string, digit: string) =>
  /^\d$/.test(digit) && icaoCheckDigit(field) === Number(digit);

export function aadhaarChecks(fields: Record<string, string>): CheckItem[] {
  const raw = (fields["aadhaar_number"] ?? "").replace(/\D/g, "");
  if (!raw) {
    return [
      {
        check: "aadhaar_format",
        label: "Number readable",
        passed: null,
        detail: "No 12-digit sequence was read. Recapture with the number in frame.",
        kind: "deterministic",
      },
    ];
  }
  const first = raw[0] ?? "0";
  return [
    {
      check: "aadhaar_format",
      label: "12-digit format",
      passed: raw.length === 12,
      detail:
        raw.length === 12 ? "Twelve digits present." : `Read ${raw.length} digits, expected 12.`,
      kind: "deterministic",
    },
    {
      check: "aadhaar_leading_digit",
      label: "Issued range",
      passed: /[2-9]/.test(first),
      detail: /[2-9]/.test(first)
        ? "Leading digit is inside the issued range (2-9)."
        : "Aadhaar numbers never begin with 0 or 1.",
      kind: "deterministic",
    },
    {
      check: "aadhaar_verhoeff",
      label: "Verhoeff check digit",
      passed: raw.length === 12 && verhoeffValid(raw),
      detail:
        raw.length === 12 && verhoeffValid(raw)
          ? "Verhoeff checksum valid — the number is internally consistent."
          : "Verhoeff checksum failed — this number cannot have been issued as printed.",
      kind: "deterministic",
    },
  ];
}

export function passportChecks(fields: Record<string, string>): CheckItem[] {
  const line2 = fields["mrz_line2"] ?? "";
  if (line2.length < 28) {
    return [
      {
        check: "passport_mrz",
        label: "Machine-readable zone",
        passed: null,
        detail: "The MRZ strip was not readable. Recapture with the two bottom lines flat and lit.",
        kind: "deterministic",
      },
    ];
  }
  const number = line2.slice(0, 9);
  const numberCd = line2.slice(9, 10);
  const dob = line2.slice(13, 19);
  const dobCd = line2.slice(19, 20);
  const expiry = line2.slice(21, 27);
  const expiryCd = line2.slice(27, 28);
  const composite = line2.slice(0, 10) + line2.slice(13, 20) + line2.slice(21, 43);
  const compositeCd = line2.slice(43, 44);

  const items: CheckItem[] = [
    {
      check: "passport_number_cd",
      label: "Passport number check digit",
      passed: icaoValid(number, numberCd),
      detail: icaoValid(number, numberCd)
        ? "ICAO 9303 check digit matches the passport number."
        : "Passport number check digit does not match (ICAO 9303).",
      kind: "deterministic",
    },
    {
      check: "passport_dob_cd",
      label: "Date of birth check digit",
      passed: icaoValid(dob, dobCd),
      detail: icaoValid(dob, dobCd)
        ? "Date of birth is consistent with its check digit."
        : "Date of birth check digit does not match.",
      kind: "deterministic",
    },
    {
      check: "passport_expiry_cd",
      label: "Expiry check digit",
      passed: icaoValid(expiry, expiryCd),
      detail: icaoValid(expiry, expiryCd)
        ? "Expiry date is consistent with its check digit."
        : "Expiry check digit does not match.",
      kind: "deterministic",
    },
  ];

  if (compositeCd) {
    items.push({
      check: "passport_composite_cd",
      label: "Composite check digit",
      passed: icaoValid(composite, compositeCd),
      detail: icaoValid(composite, compositeCd)
        ? "Composite check digit over the whole MRZ matches."
        : "Composite check digit fails — at least one MRZ field is altered or misread.",
      kind: "deterministic",
    });
  }

  const expiryDate = mrzDate(expiry);
  if (expiryDate) {
    const live = expiryDate.getTime() > Date.now();
    items.push({
      check: "passport_expiry_window",
      label: "Validity window",
      passed: live,
      detail: live
        ? `Valid until ${expiryDate.toLocaleDateString()}.`
        : `Expired on ${expiryDate.toLocaleDateString()}.`,
      kind: "deterministic",
    });
  }

  const visualPassNo = fields["visual_passport_number"];
  if (visualPassNo) {
    const mrzNo = number.replace(/</g, "");
    const match = mrzNo === visualPassNo.replace(/[^A-Z0-9]/gi, "").toUpperCase();
    items.push({
      check: "passport_mrz_vs_visual_number",
      label: "MRZ vs visual passport number",
      passed: match,
      detail: match
        ? "MRZ passport number matches the visual zone number."
        : `MRZ passport number (${mrzNo}) conflicts with visual zone (${visualPassNo}).`,
      kind: "deterministic",
    });
  }

  return items;
}

/**
 * Visa sticker (ICAO MRV-A / MRV-B machine-readable zone).
 * Same 7-3-1 arithmetic as a passport, but an MRV carries no composite digit.
 */
export function visaChecks(fields: Record<string, string>): CheckItem[] {
  const line2 = fields["mrz_line2"] ?? "";
  if (line2.length < 28) {
    return [
      {
        check: "visa_mrz",
        label: "Visa machine-readable zone",
        passed: null,
        detail:
          "The two MRZ lines on the visa sticker were not readable. Recapture them flat and lit.",
        kind: "deterministic",
      },
    ];
  }
  const number = line2.slice(0, 9);
  const numberCd = line2.slice(9, 10);
  const dob = line2.slice(13, 19);
  const dobCd = line2.slice(19, 20);
  const expiry = line2.slice(21, 27);
  const expiryCd = line2.slice(27, 28);

  const items: CheckItem[] = [
    {
      check: "visa_number_cd",
      label: "Visa number check digit",
      passed: icaoValid(number, numberCd),
      detail: icaoValid(number, numberCd)
        ? "ICAO 9303 check digit matches the visa document number."
        : "Visa number check digit does not match (ICAO 9303).",
      kind: "deterministic",
    },
    {
      check: "visa_dob_cd",
      label: "Date of birth check digit",
      passed: icaoValid(dob, dobCd),
      detail: icaoValid(dob, dobCd)
        ? "Date of birth is consistent with its check digit."
        : "Date of birth check digit does not match.",
      kind: "deterministic",
    },
    {
      check: "visa_expiry_cd",
      label: "Valid-until check digit",
      passed: icaoValid(expiry, expiryCd),
      detail: icaoValid(expiry, expiryCd)
        ? "Expiry date is consistent with its check digit."
        : "Expiry check digit does not match.",
      kind: "deterministic",
    },
  ];

  const expiryDate = mrzDate(expiry);
  if (expiryDate) {
    const live = expiryDate.getTime() > Date.now();
    items.push({
      check: "visa_expiry_window",
      label: "Visa still valid",
      passed: live,
      detail: live
        ? `Visa is valid until ${expiryDate.toLocaleDateString()}.`
        : `Visa expired on ${expiryDate.toLocaleDateString()} — overstay risk.`,
      kind: "deterministic",
    });
  }

  const holderType = (fields["mrz_line1"] ?? "").slice(0, 2);
  if (/^V/.test(holderType)) {
    items.push({
      check: "visa_mrz_type",
      label: "MRV document code",
      passed: true,
      detail: `Machine-readable visa header "${holderType}" recognised.`,
      kind: "deterministic",
    });
  }
  return items;
}

function mrzDate(yymmdd: string): Date | null {
  if (!/^\d{6}$/.test(yymmdd)) return null;
  const yy = Number(yymmdd.slice(0, 2));
  const mm = Number(yymmdd.slice(2, 4));
  const dd = Number(yymmdd.slice(4, 6));
  if (mm < 1 || mm > 12 || dd < 1 || dd > 31) return null;
  const year = yy < 70 ? 2000 + yy : 1900 + yy;
  return new Date(Date.UTC(year, mm - 1, dd));
}

const STATE_CODES = new Set([
  "AN",
  "AP",
  "AR",
  "AS",
  "BR",
  "CH",
  "CG",
  "DD",
  "DL",
  "DN",
  "GA",
  "GJ",
  "HR",
  "HP",
  "JK",
  "JH",
  "KA",
  "KL",
  "LA",
  "LD",
  "MH",
  "ML",
  "MN",
  "MP",
  "MZ",
  "NL",
  "OD",
  "OR",
  "PB",
  "PY",
  "RJ",
  "SK",
  "TN",
  "TR",
  "TS",
  "UK",
  "UA",
  "UP",
  "WB",
]);

export function dlChecks(fields: Record<string, string>): CheckItem[] {
  const raw = (fields["dl_number"] ?? "").replace(/[^A-Z0-9]/gi, "").toUpperCase();
  if (!raw) {
    return [
      {
        check: "dl_format",
        label: "Licence number readable",
        passed: null,
        detail: "No licence number pattern was read from the capture.",
        kind: "deterministic",
      },
    ];
  }
  const state = raw.slice(0, 2);
  const rto = raw.slice(2, 4);
  const year = Number(raw.slice(4, 8));
  const nowYear = new Date().getFullYear();
  return [
    {
      check: "dl_format",
      label: "RTO number grammar",
      passed: /^[A-Z]{2}\d{2}(19|20)\d{2}\d{6,7}$/.test(raw),
      detail: /^[A-Z]{2}\d{2}(19|20)\d{2}\d{6,7}$/.test(raw)
        ? "Matches the SS RR YYYY NNNNNNN national grammar."
        : "Does not match the national SS RR YYYY NNNNNNN licence grammar.",
      kind: "deterministic",
    },
    {
      check: "dl_state_code",
      label: "State code",
      passed: STATE_CODES.has(state),
      detail: STATE_CODES.has(state)
        ? `State code ${state} is a valid issuing region.`
        : `${state || "??"} is not a recognised state code.`,
      kind: "deterministic",
    },
    {
      check: "dl_rto_code",
      label: "RTO office code",
      passed: /^\d{2}$/.test(rto) && Number(rto) > 0,
      detail:
        /^\d{2}$/.test(rto) && Number(rto) > 0
          ? `RTO office code ${rto} is well-formed.`
          : "RTO office code is missing or zero.",
      kind: "deterministic",
    },
    {
      check: "dl_issue_year",
      label: "Issue year plausibility",
      passed: Number.isFinite(year) && year >= 1950 && year <= nowYear,
      detail:
        Number.isFinite(year) && year >= 1950 && year <= nowYear
          ? `Issue year ${year} is plausible.`
          : "Issue year is outside the plausible range.",
      kind: "deterministic",
    },
  ];
}

export function consistencyChecks(fields: Record<string, string>): CheckItem[] {
  const items: CheckItem[] = [];

  const dobStr = fields["dob"] || fields["date_of_birth"];
  const issueStr = fields["issue_date"] || fields["date_of_issue"];
  const expiryStr = fields["expiry_date"] || fields["date_of_expiry"] || fields["valid_until"];

  const parseDate = (d?: string): Date | null => {
    if (!d) return null;
    const match = d.match(/(\d{2})[/-](\d{2})[/-](\d{4})/);
    if (match && match[1] && match[2] && match[3]) {
      return new Date(Date.UTC(Number(match[3]), Number(match[2]) - 1, Number(match[1])));
    }
    const isoMatch = d.match(/(\d{4})[/-](\d{2})[/-](\d{2})/);
    if (isoMatch && isoMatch[1] && isoMatch[2] && isoMatch[3]) {
      return new Date(Date.UTC(Number(isoMatch[1]), Number(isoMatch[2]) - 1, Number(isoMatch[3])));
    }
    return null;
  };

  const dob = parseDate(dobStr);
  const issue = parseDate(issueStr);
  const expiry = parseDate(expiryStr);

  if (dob && issue) {
    const valid = dob.getTime() < issue.getTime();
    items.push({
      check: "consistency_dob_vs_issue",
      label: "Date of birth vs issue date",
      passed: valid,
      detail: valid
        ? "Date of birth precedes issue date."
        : "Date of birth cannot be after issue date.",
      kind: "deterministic",
    });
  }

  if (issue && expiry) {
    const valid = issue.getTime() < expiry.getTime();
    items.push({
      check: "consistency_issue_vs_expiry",
      label: "Issue date vs expiry date",
      passed: valid,
      detail: valid
        ? "Issue date precedes expiry date."
        : "Expiry date precedes or matches issue date.",
      kind: "deterministic",
    });
  }

  if (dob && expiry) {
    const valid = dob.getTime() < expiry.getTime();
    items.push({
      check: "consistency_dob_vs_expiry",
      label: "Date of birth vs expiry date",
      passed: valid,
      detail: valid
        ? "Date of birth precedes expiry date."
        : "Date of birth cannot be after or on expiry date.",
      kind: "deterministic",
    });
  }

  return items;
}

export const DISCLAIMER =
  "Structural validation is deterministic arithmetic, not AI. A pass means the number is well-formed; it does not prove the document is genuine. The officer remains the decision-maker.";
